encoder/vaapiencoder_h264.cpp

```diff
@@ -939,7 +939,10 @@ Encode_Status VaapiEncoderH264::reorder(const SurfacePtr& surface, uint64_t time
         m_reorderFrameList.push_front(picture);
         m_curFrameNum++;
         m_reorderState = VAAPI_ENC_REORD_DUMP_FRAMES;
-    }else if (m_frameIndex % (m_numBFrames + 1) != 0) {
+    } else if ((m_frameIndex % (m_numBFrames + 1) != 0)
+            && (m_frameIndex + 1 != intraPeriod())
+            && (m_frameIndex + 1 != keyFramePeriod())) {
+        // If next frame will be I or Idr, set this frame to P instead of B.
```

Comment:
As I tested, this error also happens if next frame is I frame (non-IDR). So I have to add two conditions to avoid of this issue.

Best Regards,
Xiaoxia Zhong

From: lizhong1008 [mailto:notifications@github.com]
Sent: Sunday, February 14, 2016 10:34 AM
To: 01org/libyami
Cc: Zhong, Xiaoxia
Subject: Re: [libyami] h264enc: Fix an encoding failure issue for IBBP mode. (#440)


In encoder/vaapiencoder_h264.cpp<https://github.com/01org/libyami/pull/440#discussion_r52836792>:

> @@ -939,7 +939,10 @@ Encode_Status VaapiEncoderH264::reorder(const SurfacePtr& surface, uint64_t time

>          m_reorderFrameList.push_front(picture);

>          m_curFrameNum++;

>          m_reorderState = VAAPI_ENC_REORD_DUMP_FRAMES;

> -    }else if (m_frameIndex % (m_numBFrames + 1) != 0) {

> +    } else if ((m_frameIndex % (m_numBFrames + 1) != 0)

> +            && (m_frameIndex + 1 != intraPeriod())

> +            && (m_frameIndex + 1 != keyFramePeriod())) {

> +        // If next frame will be I or Idr, set this frame to P instead of B.

Only when next frame is IDR, need to set this frame to be P frame, because only IDR( not I frame ) need to refresh DPB buffer.

—
Reply to this email directly or view it on GitHub<https://github.com/01org/libyami/pull/440/files#r52836792>.